teachopencadd/talktorials/T018_automated_cadd_pipeline/utils/helpers/pubchem.py

```diff
@@ -0,0 +1,298 @@
+        + str(input_id_value)
+        + getattr(APIConsts.URLs.Operations, f"GET_{output_id_type}".upper()).value
+    )
+    response_data = send_request(url, output_data_type)
+
+    if isinstance(input_id_value, list):
+        response_data = response_data.strip().split("\n")
+    else:
+        response_data = response_data.strip()
+    return response_data
+
+
+def get_compound_record(input_id_type, input_id_value, output_data_type="json"):
+    """
+    Get a full record of all physiochemical properties of the compound.
+
+    Parameters
+    ----------
+    input_id_type : str
+        Type of the input identifier.
+        Valid values are: 'name', 'cid', 'smiles', 'inchi' and 'inchikey'.
+        Valid values are stored in: `APIConsts.URLs.Inputs`
+    input_id_value : str or int
+        Value of the input identifier.
+    output_data_type : str
+        Optional; default: 'txt'.
+        Datatype of the output data.
+        Valid values are 'txt', 'json', 'csv'.
+        A list of all valid values are stored in: `APIConsts.URLs.Outputs`
+
+    Returns
+    -------
+    dict
+        Dictionary keys are: 'id', 'atoms', 'bonds', 'coords', 'charge', 'props', 'count'
+    """
+    url = (
+        getattr(APIConsts.URLs.Inputs, input_id_type.upper()).value
+        + str(input_id_value)
+        + getattr(APIConsts.URLs.Operations, "GET_RECORD").value
+    )
+    response_data = send_request(url, output_data_type)[
+        APIConsts.ResponseMsgs.GetRecords.RESPONSE_KEY.value
+    ][0]
+
+    return response_data
+
+
+def get_description_from_smiles(smiles, output_data_type="json", printout=False):
+    """
+    Get a textual description of a molecule, including its usage, properties, source etc.
+
+    Parameters
+    ----------
+    smiles : str
+        SMILES of the molecule.
+    output_data_type : str
+        Optional; default: 'txt'.
+        Datatype of the output data.
+        Valid values are 'txt', 'json', 'csv'.
+        A list of all valid values are stored in: `APIConsts.URLs.Outputs`
+    printout : bool
+        Whether to print the descriptions, or return the data.
+
+    Returns
+    -------
+    list of dict
+        When the parameter printout is set to False, the raw data is
+        returned as a list of dicts, where each element of the list
+        corresponds to a description from a specific source (e.g. a journal article).
+    """
+    url = (
+        APIConsts.URLs.Inputs.SMILES.value
+        + smiles
+        + APIConsts.URLs.Operations.GET_DESCRIPTION.value
+    )
+    response_data = send_request(url, output_data_type)[
+        APIConsts.ResponseMsgs.GetDescription.RESPONSE_KEY1.value
+    ][APIConsts.ResponseMsgs.GetDescription.RESPONSE_KEY2.value]
+
+    if printout:
+        for entry in response_data:
+            try:
+                print(entry["Description"] + "\n")
+            except KeyError:
+                print(entry)
+    else:
+        return response_data
```

Comment:
@Armin-Ariamajd, can you please comment on each line of code here? (The `except KeyError` is my doing I think, might need updating.)

Reply:
Again, similar to the case in the `dogsitescorer.py` file I just explained, this `try-except` block is also not supposed to really catch errors, but is supposed to jump over some list entries:

This is in a function to return a textual description of a compound by providing its smiles. The PubChem query returns a list of dictionaries like this:
```
[{'CID': 2244, 'Title': 'Aspirin'},
 {'CID': 2244,
  'Description': 'Acetylsalicylic acid is a member of the class of benzoic acids that is salicylic acid in which the hydrogen that is attached to the phenolic hydroxy group has been replaced by an acetoxy group. A non-steroidal anti-inflammatory drug with cyclooxygenase inhibitor activity. It has a role as a non-steroidal anti-inflammatory drug, a non-narcotic analgesic, a platelet aggregation inhibitor, an antipyretic, a cyclooxygenase 2 inhibitor, a cyclooxygenase 1 inhibitor, a prostaglandin antagonist, a teratogenic agent, an anticoagulant, a plant activator, an EC 1.1.1.188 (prostaglandin-F synthase) inhibitor, a drug allergen and a geroprotector. It is a member of benzoic acids, a member of salicylates and a member of phenyl acetates. It derives from a salicylic acid. It is a conjugate acid of an acetylsalicylate.',
  'DescriptionSourceName': 'ChEBI',
  'DescriptionURL': 'http://www.ebi.ac.uk/chebi/searchId.do?chebiId=CHEBI:15365'}] 
```

if the ```printout``` parameter is set to ```False```, the function just returns this list of dictionaries. But if it's set to `True`, then it is supposed to only print the actual description to the console. So it iterates over the list (where each element of the list is a dictionary) and looks for the dictionary (or dictionaries, since some compounds have multiple descriptions from different sources) which has a key `Description`. It then prints those descriptions to the console. The `except: pass` is to ignore the dictionaries that don't have the `Description` key, like the first element of the list in the above example.

Reply:
`try/except` clauses for dictionary access are usually circumvented with `dict.get()` instead.

```python
print(entry.get("Description", entry))
```

Reply:
Thanks @jaimergp, I didn't know that. However, we don't want to print the entire dict if it doesn't contain the key `Description`. On the other hand, looking at a number of responses for various compounds, it seems that they always follow the same pattern, namely that the first dictionary always only contains `Title` and `CID`, and the next dictionaries in the list (if any), each contain `Description`, `DescriptionSourceName` and `DescriptionURL`. So I changed the code to this (see 735b418):
 
```python
if printout:
        print(
            f"Name: {response_data[0][APIConsts.ResponseMsgs.GetDescription.TITLE_KEY.value]}\n"
            f"CID: {response_data[0][APIConsts.ResponseMsgs.GetDescription.CID_KEY.value]}\n"
        )
        if len(response_data) == 1:
            print("No description available.")
        else:
            for entry in response_data[1:]:
                print(
                    f"{entry[APIConsts.ResponseMsgs.GetDescription.DESCRIPTION_KEY.value]}\n"
                    f"Source: {entry[APIConsts.ResponseMsgs.GetDescription.DESCRIPTION_SOURCE_NAME_KEY.value]}, {entry[APIConsts.ResponseMsgs.GetDescription.DESCRIPTION_SOURCE_URL.value]}\n\n"
                )
    else:
        return response_data
```
which prints something like this:
```
Name: 2-Amino-3,4-dimethylimidazo(4,5-F)quinoline
CID: 62274

MeIQ is an aminoquinoline.
Source: ChEBI, http://www.ebi.ac.uk/chebi/searchId.do?chebiId=CHEBI:82338
```